library/core/src/iter/range/ptr.rs

```diff
@@ -0,0 +1,274 @@
+                    // nonempty range.
+                    (usize::MAX, None)
+                } else {
+                    // In between self.start and self.end there are some number
+                    // of whole elements of type T, followed by possibly a
+                    // remainder element if T's size doesn't evenly divide the
+                    // byte distance between the endpoints. The remainder
+                    // element still counts as being part of this range, since
+                    // the pointer to it does lie between self.start and
+                    // self.end.
+                    let byte_offset = self.end as usize - self.start as usize;
+                    let number_of_whole_t = byte_offset / mem::size_of::<T>();
+                    let remainder_bytes = byte_offset % mem::size_of::<T>();
+                    let maybe_remainder_t = (remainder_bytes > 0) as usize;
+                    let hint = number_of_whole_t + maybe_remainder_t;
+                    (hint, Some(hint))
+                }
+            }
+
+            fn nth(&mut self, n: usize) -> Option<Self::Item> {
+                let _ = self.advance_by(n);
+                self.next()
+            }
+
+            fn last(mut self) -> Option<Self::Item> {
+                self.next_back()
+            }
+
+            fn min(mut self) -> Option<Self::Item> {
+                self.next()
+            }
+
+            fn max(mut self) -> Option<Self::Item> {
+                self.next_back()
+            }
+
+            fn is_sorted(self) -> bool {
+                true
+            }
+
+            fn advance_by(&mut self, n: usize) -> Result<(), usize> {
+                match self.size_hint().1 {
+                    None => {
+                        // T is zero sized. Advancing does nothing.
+                        Ok(())
+                    }
+                    Some(len) => match n.cmp(&len) {
+                        Ordering::Less => {
+                            // Advance past n number of whole elements.
+                            self.start = self.start.wrapping_add(n);
+                            Ok(())
+                        }
+                        Ordering::Equal => {
+                            // Advance past every single element in the
+                            // iterator, including perhaps the remainder
+                            // element, leaving an empty iterator.
+                            self.start = self.end;
+                            Ok(())
+                        }
+                        Ordering::Greater => {
+                            // Advance too far.
+                            self.start = self.end;
+                            Err(len)
+                        }
+                    }
+                }
+            }
+
+            #[doc(hidden)]
+            unsafe fn __iterator_get_unchecked(&mut self, idx: usize) -> Self::Item {
+                self.start.wrapping_add(idx)
```

Comment:
Since `TrustedRandomAccess` requires that the passed `idx` always is in bounds souldn't `.add()` be correct here?